FE-1162 FE-1165 Use localized thousands and decimals separator

PavlosTze opened the pull request:
### **Why?**
For example 1000,2 should be shown as 1.000,2 (and respect user's locale in each case)

### **How?**
- Added `decimalFormat.groupingSize = 3` and `decimalFormat.isGroupingUsed = true` in `formatTokens` function
- Replaced `NumberFormat.getInstance()` with `NumberFormat.getInstance(Locale.getDefault())`
- Used a localized decimal separator in `CompactDecimalFormat`
- Moved `EMPTY_VALUE` to `Contracts.kt` and `formatTokens` functions in `NumberUtils`.
- Use Proper Localization in `Weather` functions as we now use `formatNumber` functionality and not directly transforming `Float` values to `String`.
- Use DI for `DecimalFormat` as an optimization instead of reinitializing it every time `formatTokens` is called

### **Testing**
- Ensure that all "WXM-related" numbers look OK (and localized when it comes to thousands and decimal separators)
- Ensure that in Network Stats screen the station stats at the bottom of the screen have correct localized numbers
- Ensure that all Weather-related numbers look OK (and localized when it comes to thousands and decimal separators) 
 <div id='description'>
<h3>Summary by Bito</h3>
This pull request implements localized number formatting throughout the WeatherXM app, refactoring utility functions in NumberUtils and Weather classes. It updates UI components, charts, and rewards/stats use cases to use locale-specific formatting. The changes include improved dependency injection for DecimalFormat and extensive updates to test cases for various locales.
<br>
<br>
<b>Code change type</b>: Refactoring, Optimization, Internationalization
<br>
<br>
<b>Unit tests added</b>: True
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 5
</div>

PavlosTze commented:
@CodiumAI-Agent /review

CodiumAI-Agent commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<a href='https://github.com/WeatherXM/wxm-android/pull/342/files#diff-bd426c1e06513b3624fbbdb4369a298b97a6b86fc53e4599051df64250860b12R843-R862'><strong>Localization Consistency</strong></a><br>Ensure that the localization settings (e.g., decimal and thousands separators) are consistently applied across different locales. The current implementation may not handle locale changes dynamically, which could lead to inconsistencies in displayed data.

<a href='https://github.com/WeatherXM/wxm-android/pull/342/files#diff-93d573514ac74bc25440a5e55362e5ccc0925e619e04053f2755d3735e9132b4R36-R59'><strong>Decimal Formatting</strong></a><br>The use of `DecimalFormat` for formatting numbers might not be consistent or might not handle edge cases well. Review and ensure that the formatting is appropriate for all expected input ranges and locales.

</td></tr>
</table>

pantelisss reviewed (approved):
🤌

PavlosTze commented:
/review

PavlosTze commented:
AI Code Review is in progress (usually takes <3 minutes unless it's a very large PR).

CodiumAI-Agent commented:
**[Persistent review](https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388026410)** updated to latest commit https://github.com/WeatherXM/wxm-android/commit/5f0ff10bd6db6574ab7b3b363149a08ec8a80084

PavlosTze commented:
<div>
<h2>Changelist by <a href="https://bito.ai">Bito</a></h2>
<p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>Feature Improvement</b> - Localized Number Formatting
</td>
<td width="70%">


<div>

-&nbsp;[`Modules.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-bd426c1e06513b3624fbbdb4369a298b97a6b86fc53e4599051df64250860b12) - Added localized decimal formatting and thousands grouping
<div>

-&nbsp;[`NumberUtils.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-93d573514ac74bc25440a5e55362e5ccc0925e619e04053f2755d3735e9132b4) - Implemented localized number formatting and moved token formatting functions
<div>

-&nbsp;[`Weather.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-6459b6951b6d575758eb81c642de63115fbbd9567d707f5bba9febbd254ecd5f) - Updated to use new localized number formatting
<div>

-&nbsp;[`Rewards.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-dea2928dba0d0f0147b9a9514bfcdd016d90fccbdb06adeb4173b22ae5bb120a) - Removed redundant formatting functions
<div>

-&nbsp;[`Contracts.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-99dc566e4fc47492250324ac735658bc42bfec4c4eb907628d4d427145baea7a) - Added EMPTY_VALUE constant
<div>

-&nbsp;[`RewardsUseCase.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-c020e74fa6d878405cc5f489f4aaec3f54c5af7e7a608c5ac53f1718575241b1) - Updated to use new number formatting
<div>

-&nbsp;[`StatsUseCaseImpl.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-85ccc38aa5032ccb75344dc7aafdafb1fd4eb35977028733f26f36d58e294d11) - Updated to use new EMPTY_VALUE constant
<div>

-&nbsp;[`ChartsAxisFormatters.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-4aaaec63d43078f8e2721d000e99bcb741d5d62b2578837c8ff1dcf71db8c824) - Updated to use new number formatting
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>Other Improvements</b> - Code Reorganization and Dependency Injection
</td>
<td width="70%">


<div>

-&nbsp;[`RewardSplitStakeholderAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-f61aedbe827811c3e4016a82a87d8292b7aaf3166b0c487984d660278bc5a30d) - Updated import for formatTokens function
<div>

-&nbsp;[`DailyRewardsCardView.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-3574139a5002aaef6d084c8965dd0cb91cc0a148f3aad46d792f2f8f287f65e3) - Updated imports and simplified reward formatting
<div>

-&nbsp;[`TooltipMarkerView.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-a6293bfe34edbba50bd2a6c8c1eb6a48877eb1c7bc4823c029fab071f49c5d4b) - Updated import for formatTokens function
<div>

-&nbsp;[`RewardsFragment.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-c59a302e816df5da7ff9e709216d4aefc403a513cb2759d53a478aa234c1ed53) - Updated imports and formatting function usage
<div>

-&nbsp;[`DeviceRewardsAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-27018e36e02bb34ab7beadae4fc490e75fc575b15f00d2f34682ff5f9064c8f0) - Updated import for formatTokens function
<div>

-&nbsp;[`DeviceRewardsBoostAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-9f143a7bd9703fd76c04ebf8964c5054c75f2262b5b813f64a75bdae3b5d1160) - Updated import for formatTokens function
<div>

-&nbsp;[`DevicesRewardsActivity.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-afe6c32db1e251082c2860be7157467793fa09f531f1e973aa5e69fdbdb9710c) - Updated imports and formatting function usage
<div>

-&nbsp;[`DevicesFragment.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-a52290c1282d8f5b84878e0a5d8a6011e799f5b2bd6c488b3803eea3fcbf3a01) - Updated imports and formatting function usage
<div>

-&nbsp;[`ProfileFragment.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-b926f24b0a55ddb291ecb0574f386628bab1a9f2c6f31e7c4b8b1cdb9c2e9534) - Updated imports and formatting function usage
<div>

-&nbsp;[`RewardBoostAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-be3c6a0b06e70763ce65f278ecd6618bb418065a8e486028ee0a370d32c833a7) - Updated import for formatTokens function
<div>

-&nbsp;[`RewardDetailsActivity.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-042d7bd3ff1716d0e9c6fae15bdc42bd8be631587668bd81c94b3c5027327324) - Updated imports and formatting function usage
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>Testing</b> - Test Updates for Localized Formatting
</td>
<td width="70%">


<div>

-&nbsp;[`RewardsUseCaseTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-4663f498f776eacc5bfd5d8e0e15274766b8f7def845a5a01086c478ac4e09b0) - Updated test setup for localized number formatting
<div>

-&nbsp;[`StatsUseCaseTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-46f41333424380062399b9fbadb3d1f86fd3d3c87bb8c5302fffe8d8cd4f8c4b) - Updated test cases for localized number formatting
<div>

-&nbsp;[`ChartsAxisFormattersTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-c162a2db1e48b1e4642ba5798c7e74a90bdb708bd06bb68209f0da26b67f5fb8) - Added Koin setup for localized number formatting tests
<div>

-&nbsp;[`NumberUtilsTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-5bf0d35d60bd09c9661de1d71d2563934fbd1283a219060c8e2026d73a676dc8) - Updated imports and test cases for new number formatting
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>Feature Improvement</b> - Localized Number Formatting Enhancement
</td>
<td width="70%">


<div>

-&nbsp;[`NumberUtilsTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-5bf0d35d60bd09c9661de1d71d2563934fbd1283a219060c8e2026d73a676dc8) - Updated test cases for localized number formatting and added new tests for formatting tokens
<div>

-&nbsp;[`RewardsTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-ca81bb2eeda386f7ffedad3fd9c3dfe6e4f581773525a81708185b7b7cd00c80) - Removed redundant tests for token formatting
<div>

-&nbsp;[`WeatherExtensionsTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-8bfaec2dcf1caab00cfdec727118c6eb5388ef298aef49ae59bbf08463a12503) - Updated imports and test cases for localized pressure formatting
<div>

-&nbsp;[`WeatherTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-bc175dc440950bf56d0c0b7571a3869149f69c6061090e7bfbf3467c72dae0a1) - Added localized number formatting setup and updated pressure tests
<div>

-&nbsp;[`Modules.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-bd426c1e06513b3624fbbdb4369a298b97a6b86fc53e4599051df64250860b12) - Added localized decimal formatting and thousands grouping
<div>

-&nbsp;[`NumberUtils.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-93d573514ac74bc25440a5e55362e5ccc0925e619e04053f2755d3735e9132b4) - Implemented localized number formatting and moved token formatting functions
<div>

-&nbsp;[`Weather.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-6459b6951b6d575758eb81c642de63115fbbd9567d707f5bba9febbd254ecd5f) - Updated to use new localized number formatting
<div>

-&nbsp;[`Rewards.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-dea2928dba0d0f0147b9a9514bfcdd016d90fccbdb06adeb4173b22ae5bb120a) - Removed redundant formatting functions
<div>

-&nbsp;[`Contracts.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-99dc566e4fc47492250324ac735658bc42bfec4c4eb907628d4d427145baea7a) - Added EMPTY_VALUE constant
<div>

-&nbsp;[`RewardsUseCase.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-c020e74fa6d878405cc5f489f4aaec3f54c5af7e7a608c5ac53f1718575241b1) - Updated to use new number formatting
<div>

-&nbsp;[`StatsUseCaseImpl.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-85ccc38aa5032ccb75344dc7aafdafb1fd4eb35977028733f26f36d58e294d11) - Updated to use new EMPTY_VALUE constant
<div>

-&nbsp;[`ChartsAxisFormatters.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-4aaaec63d43078f8e2721d000e99bcb741d5d62b2578837c8ff1dcf71db8c824) - Updated to use new number formatting
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>Other Improvements</b> - Code Reorganization and Dependency Injection
</td>
<td width="70%">


<div>

-&nbsp;[`RewardSplitStakeholderAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-f61aedbe827811c3e4016a82a87d8292b7aaf3166b0c487984d660278bc5a30d) - Updated import for formatTokens function
<div>

-&nbsp;[`DailyRewardsCardView.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-3574139a5002aaef6d084c8965dd0cb91cc0a148f3aad46d792f2f8f287f65e3) - Updated imports and simplified reward formatting
<div>

-&nbsp;[`TooltipMarkerView.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-a6293bfe34edbba50bd2a6c8c1eb6a48877eb1c7bc4823c029fab071f49c5d4b) - Updated import for formatTokens function
<div>

-&nbsp;[`RewardsFragment.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-c59a302e816df5da7ff9e709216d4aefc403a513cb2759d53a478aa234c1ed53) - Updated imports and formatting function usage
<div>

-&nbsp;[`DeviceRewardsAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-27018e36e02bb34ab7beadae4fc490e75fc575b15f00d2f34682ff5f9064c8f0) - Updated import for formatTokens function
<div>

-&nbsp;[`DeviceRewardsBoostAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-9f143a7bd9703fd76c04ebf8964c5054c75f2262b5b813f64a75bdae3b5d1160) - Updated import for formatTokens function
<div>

-&nbsp;[`DevicesRewardsActivity.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-afe6c32db1e251082c2860be7157467793fa09f531f1e973aa5e69fdbdb9710c) - Updated imports and formatting function usage
<div>

-&nbsp;[`DevicesFragment.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-a52290c1282d8f5b84878e0a5d8a6011e799f5b2bd6c488b3803eea3fcbf3a01) - Updated imports and formatting function usage
<div>

-&nbsp;[`ProfileFragment.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-b926f24b0a55ddb291ecb0574f386628bab1a9f2c6f31e7c4b8b1cdb9c2e9534) - Updated imports and formatting function usage
<div>

-&nbsp;[`RewardBoostAdapter.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-be3c6a0b06e70763ce65f278ecd6618bb418065a8e486028ee0a370d32c833a7) - Updated import for formatTokens function
<div>

-&nbsp;[`RewardDetailsActivity.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-042d7bd3ff1716d0e9c6fae15bdc42bd8be631587668bd81c94b3c5027327324) - Updated imports and formatting function usage
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>Testing</b> - Test Updates for Localized Formatting
</td>
<td width="70%">


<div>

-&nbsp;[`RewardsUseCaseTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-4663f498f776eacc5bfd5d8e0e15274766b8f7def845a5a01086c478ac4e09b0) - Updated test setup for localized number formatting
<div>

-&nbsp;[`StatsUseCaseTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-46f41333424380062399b9fbadb3d1f86fd3d3c87bb8c5302fffe8d8cd4f8c4b) - Updated test cases for localized number formatting
<div>

-&nbsp;[`ChartsAxisFormattersTest.kt`](https://github.com/WeatherXM/wxm-android/pull/342/files#diff-c162a2db1e48b1e4642ba5798c7e74a90bdb708bd06bb68209f0da26b67f5fb8) - Added Koin setup for localized number formatting tests
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

PavlosTze commented on app/src/test/java/com/weatherxm/util/WeatherTest.kt:
```diff
@@ -149,11 +154,11 @@ class WeatherTest : KoinTest, BehaviorSpec({
                 and("unit is in inHg") {
                     every { sharedPref.getString("key_pressure_preference", "hPa") } returns "inHg"
-                    testPressure("inHg", 29.54F)
+                    testPressure("inHg", "29.54")
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case doesn't explicitly verify the localization of the pressure value, which is a key aspect of the changes described in the PR description.
</div>
<div id="fix">
<b>Fix:</b> Modify the test case to explicitly check for correct localization of the pressure value. This should include tests with different locales to ensure proper formatting of decimal and thousands separators.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -160,3 +160,9 @@
      every { sharedPref.getString("key_pressure_preference", "hPa") } returns "inHg"
 -    testPressure("inHg", "29.54")
 +    testPressureLocalization("inHg", 29.54f, Locale.US, "29.54")
 +    testPressureLocalization("inHg", 29.54f, Locale.GERMANY, "29,54")
 +}
 +and("unit is in hPa with high value") {
 +    every { sharedPref.getString("key_pressure_preference", "hPa") } returns "hPa"
 +    testPressureLocalization("hPa", 1000.54f, Locale.US, "1,000.54")
 +    testPressureLocalization("hPa", 1000.54f, Locale.GERMANY, "1.000,54")
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/test/java/com/weatherxm/util/NumberUtilsTest.kt:
```diff
@@ -20,13 +29,19 @@ class NumberUtilsTest : BehaviorSpec({
                     }
                     single<NumberFormat> {
-                        mockk<NumberFormat>()
+                        NumberFormat.getInstance(Locale.US)
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> Using a fixed Locale (US) for NumberFormat instance may not be suitable for internationalization and localization requirements of the app
</div>
<div id="fix">
<b>Fix:</b> Replace the fixed Locale.US with Locale.getDefault() to use the device's locale settings, or inject the appropriate locale based on the app's localization strategy
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -32 +32 @@
 -                        NumberFormat.getInstance(Locale.US)
 +                        NumberFormat.getInstance(Locale.getDefault())
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/test/java/com/weatherxm/util/ChartsAxisFormattersTest.kt:
```diff
@@ -2,13 +2,30 @@ package com.weatherxm.util
+                }
+            )
+        }
+    }
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The Koin context is initialized but not properly cleaned up after the tests. A similar issue was also found in app/src/test/java/com/weatherxm/util/ChartsAxisFormattersTest.kt (line 28).
</div>
<div id="fix">
<b>Fix:</b> Add an afterSpec block to stop the Koin context after all tests have completed.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -27,6 +27,10 @@ class ChartsAxisFormattersTest : BehaviorSpec({
      }
  }

 +    afterSpec {
 +        stopKoin()
 +    }
 +
  context("Use custom X axis formatter") {
      given("A list of times") {
          then("it should return the correct time") {
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/test/java/com/weatherxm/util/ChartsAxisFormattersTest.kt:
```diff
@@ -24,12 +41,12 @@ class ChartsAxisFormattersTest : BehaviorSpec({
                 and("The formatter supports decimals") {
                     then("it should return the correct value") {
-                        yAxisFormatterWithDecimals.getAxisLabel(1000.65F, null) shouldBe "1000.7"
+                        yAxisFormatterWithDecimals.getAxisLabel(1000.65F, null) shouldBe "1,000.7"
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case assumes a specific locale (likely US) for the thousands separator, which may not be consistent across all locales.
</div>
<div id="fix">
<b>Fix:</b> Update the test case to use a specific locale for testing, and add additional test cases for different locales to ensure consistent behavior across various number formatting conventions.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -46,7 +46,15 @@
              }
              and("The formatter supports decimals") {
                  then("it should return the correct value") {
 -                    yAxisFormatterWithDecimals.getAxisLabel(1000.65F, null) shouldBe "1,000.7"
 +                    withLocale(Locale.US) {
 +                        yAxisFormatterWithDecimals.getAxisLabel(1000.65F, null) shouldBe "1,000.7"
 +                    }
 +                    withLocale(Locale.GERMANY) {
 +                        yAxisFormatterWithDecimals.getAxisLabel(1000.65F, null) shouldBe "1.000,7"
 +                    }
 +                    withLocale(Locale.FRANCE) {
 +                        yAxisFormatterWithDecimals.getAxisLabel(1000.65F, null) shouldBe "1 000,7"
 +                    }
                  }
              }
          }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/test/java/com/weatherxm/util/ChartsAxisFormattersTest.kt:
```diff
@@ -40,4 +57,8 @@ class ChartsAxisFormattersTest : BehaviorSpec({
         }
     }
+
+    afterSpec {
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The afterSpec block is added to clean up Koin resources, but there's no corresponding setup visible in the provided context. This may lead to inconsistent test states or errors if Koin is not properly initialized before the tests run.
</div>
<div id="fix">
<b>Fix:</b> Add a corresponding Koin initialization in a beforeSpec block to ensure proper setup before running tests. Also, consider moving Koin setup and teardown to individual test cases if fine-grained control is needed.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -11,6 +11,13 @@ import java.util.Locale

  class ChartsAxisFormattersTest : BehaviorSpec({

 +    beforeSpec {
 +        startKoin {
 +            modules(module {
 +                // Add necessary Koin configurations for the test
 +            })
 +        }
 +    }
 +
      // ... existing test code ...

      afterSpec {
          stopKoin()
      }
  })
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/ChartsAxisFormatters.kt:
```diff
@@ -20,13 +19,11 @@ class CustomYAxisFormatter(
-        } else if (value < 10000) {
-            "${roundToInt(value)}"
+        val label = if (value < 10000) {
+            formatNumber(value, decimals)
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The change from roundToDecimals to formatNumber may alter the precision and formatting of the axis labels, potentially affecting the chart's readability and accuracy.
</div>
<div id="fix">
<b>Fix:</b> Add the 'enforceDecimals = true' parameter to the formatNumber function call to ensure that the specified number of decimal places is always used, maintaining the same precision as the previous roundToDecimals function.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -23,2 +23,2 @@
 -            formatNumber(value, decimals)
 +            formatNumber(value, decimals, enforceDecimals = true)
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/NumberUtils.kt:
```diff
@@ -1,35 +1,61 @@
 
-    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
-        return value.toFloat().toBigDecimal().setScale(decimals, RoundingMode.HALF_UP).toFloat()
+    fun formatTokens(amount: Float?): String {
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The formatTokens function for Float may introduce rounding errors when converting to BigDecimal, which could lead to inaccuracies in financial calculations.
</div>
<div id="fix">
<b>Fix:</b> Use BigDecimal for the input parameter instead of Float to maintain precision. If Float must be used, consider using BigDecimal's valueOf method for more accurate conversion.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -36,7 +36,7 @@ object NumberUtils : KoinComponent {
 -    fun formatTokens(amount: Float?): String {
 -        return amount?.let {
 -            formatTokens(it.toBigDecimal())
 +    fun formatTokens(amount: Float?): String {
 +        return amount?.let {
 +            formatTokens(BigDecimal.valueOf(it.toDouble()))
      } ?: EMPTY_VALUE
  }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/NumberUtils.kt:
```diff
@@ -1,35 +1,61 @@
+        return decimalFormat.format(amount)
+    }
+
+    @Suppress("MagicNumber")
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2388448877>#04ac52</a><span> - 10/02/2024, 12:12 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The weiToETH function doesn't handle potential overflow for very large Wei amounts, which could lead to incorrect ETH calculations or runtime exceptions.
</div>
<div id="fix">
<b>Fix:</b> Implement proper overflow checking and handling in the weiToETH function. Consider using BigInteger for intermediate calculations to avoid potential overflow.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -46,10 +46,15 @@ object NumberUtils : KoinComponent {
  @Suppress("MagicNumber")
  fun weiToETH(amount: BigDecimal): BigDecimal {
      /**
       * Mandatory otherwise we get a result of 0E-18 instead of 0
       */
      if (amount == BigDecimal.ZERO) {
          return BigDecimal.ZERO
      }
 -    return amount.divide(BigDecimal(DIVISOR_WEI_TO_ETH), ETH_DECIMALS, RoundingMode.HALF_UP)
 +    val divisor = BigDecimal(DIVISOR_WEI_TO_ETH)
 +    return try {
 +        amount.divide(divisor, ETH_DECIMALS, RoundingMode.HALF_UP)
 +    } catch (e: ArithmeticException) {
 +        Log.e("NumberUtils", "Error converting Wei to ETH: ${e.message}")
 +        BigDecimal.ZERO // or handle the error appropriately
 +    }
  }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/NumberUtils.kt:
```diff
@@ -1,35 +1,61 @@
     private val numberFormat: NumberFormat by inject()
+    private val decimalFormat: DecimalFormat by inject(named(DECIMAL_FORMAT_TOKENS))
+
+    private const val DIVISOR_WEI_TO_ETH = "1000000000000000000"
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2391143597>#b921d3</a><span> - 10/03/2024, 11:34 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The DIVISOR_WEI_TO_ETH constant is defined as a string, which may lead to inefficient operations and potential loss of precision when used in calculations involving Wei to ETH conversions.
</div>
<div id="fix">
<b>Fix:</b> Replace the string constant with a BigInteger or BigDecimal constant for DIVISOR_WEI_TO_ETH to ensure precise calculations when converting Wei to ETH.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -19,2 +19,2 @@ object NumberUtils : KoinComponent {
 -    private const val DIVISOR_WEI_TO_ETH = "1000000000000000000"
 +    private val DIVISOR_WEI_TO_ETH = BigInteger("1000000000000000000")
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/NumberUtils.kt:
```diff
@@ -1,35 +1,61 @@
 
-    fun roundToInt(value: Number): Int {
-        return value.toFloat().toBigDecimal().setScale(0, RoundingMode.HALF_UP).toInt()
+    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2391143597>#b921d3</a><span> - 10/03/2024, 11:34 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The roundToDecimals function lacks input validation for the 'decimals' parameter, which could lead to unexpected behavior or errors if negative or excessively large values are provided.
</div>
<div id="fix">
<b>Fix:</b> Add input validation for the 'decimals' parameter to ensure it's non-negative and within a reasonable range.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -57,2 +57,7 @@ object NumberUtils : KoinComponent {
 -    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
 -        return value.toFloat().toBigDecimal().setScale(decimals, RoundingMode.HALF_UP).toFloat()
 +    fun roundToDecimals(value: Number, decimals: Int = 1): Float {
 +        require(decimals >= 0) { "Decimals must be non-negative" }
 +        require(decimals <= 10) { "Decimals must be 10 or less" }
 +        return value.toFloat().toBigDecimal().setScale(decimals, RoundingMode.HALF_UP).toFloat()
 +    }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/Weather.kt:
```diff
@@ -106,15 +106,13 @@ object Weather : KoinComponent {
-            } else {
-                "${roundToDecimals(value)}$unit"
-            }
+        val formattedValue = if (ignoreConversion) {
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2391143597>#b921d3</a><span> - 10/03/2024, 11:34 am</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The function uses multiple boolean parameters, which can make the code less readable and more prone to errors when calling the function.
</div>
<div id="fix">
<b>Fix:</b> Consider using a data class or a builder pattern to replace the multiple boolean parameters. This will make the function calls more readable and less error-prone.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -109,13 +109,26 @@ object Weather : KoinComponent {
 -    fun getFormattedTemperature(
 -        value: Float?,
 -        decimals: Int = 0,
 -        fullUnit: Boolean = true,
 -        includeUnit: Boolean = true,
 -        ignoreConversion: Boolean = false
 -    ): String {
 -        val formattedValue = if (ignoreConversion) {
 -            formatNumber(value, decimals)
 -        } else {
 -            formatNumber(convertTemp(value, decimals), decimals)
 -        }
 -
 -        return "$formattedValue$unit"
 +    data class TemperatureFormatOptions(
 +        val decimals: Int = 0,
 +        val fullUnit: Boolean = true,
 +        val includeUnit: Boolean = true,
 +        val ignoreConversion: Boolean = false
 +    )
 +
 +    fun getFormattedTemperature(
 +        value: Float?,
 +        options: TemperatureFormatOptions = TemperatureFormatOptions()
 +    ): String {
 +        val (decimals, fullUnit, includeUnit, ignoreConversion) = options
 +        val formattedValue = if (ignoreConversion) {
 +            formatNumber(value, decimals)
 +        } else {
 +            formatNumber(convertTemp(value, decimals), decimals)
 +        }
 +
 +        val unit = if (includeUnit) {
 +            if (fullUnit) "°${getTemperatureUnit()}" else "°"
 +        } else ""
 +
 +        return "$formattedValue$unit"
 +    }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

PavlosTze commented on app/src/main/java/com/weatherxm/util/Weather.kt:
```diff
@@ -106,15 +106,13 @@ object Weather : KoinComponent {
-                "${roundToDecimals(value)}$unit"
-            }
+        val formattedValue = if (ignoreConversion) {
+            formatNumber(value, decimals)
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/WeatherXM/wxm-android/pull/342#issuecomment-2391143597>#b921d3</a><span> - 10/03/2024, 11:34 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The formatNumber function is used without considering the possibility of null return values, which could lead to potential null pointer exceptions.
</div>
<div id="fix">
<b>Fix:</b> Use the Elvis operator (?:) to provide a default value in case formatNumber returns null. This ensures that a non-null value is always used.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -109,7 +109,7 @@ object Weather : KoinComponent {
      }

      val formattedValue = if (ignoreConversion) {
 -        formatNumber(value, decimals)
 +        formatNumber(value, decimals) ?: EMPTY_VALUE
      } else {
 -        formatNumber(convertTemp(value, decimals), decimals)
 +        formatNumber(convertTemp(value, decimals), decimals) ?: EMPTY_VALUE
      }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged
